### ISSUE TITLE
Cannot encode object: <object>

### ISSUE THREAD

**Issue by enugentdt** [Scille/umongo#169]
Hi there,

So, I've found an "issue." Not really an issue, but an interesting feature request caused by an exception, shall we say? Let's say you have an object, Foo, which contains a ReferenceField to Bar. If you search for the Foo object (Foo.find({...})) with the specific reference to Bar, passed as a single ObjectID, it works. However, if you run a find, and pass in an _entire object_ as the parameter, it fails, with a very long error message, along the lines of `Cannot encode object: <object>`.

What I propose is to detect if the user is searching for a ReferenceField, and they passed an object. If they passed an object, simply obtain the `pk` field, and use that instead.

I could implement this on my own servers easily, however, I feel this would be a good addition to the database. 

To be clear, this _specific_ fix is just one option of many, and can go in a lot of places. But here's a possibility:

`umongo/query_mapper.py` line `41`
```python
    for entry, entry_query in query.items():
        if isinstance(entry_query, umongo.DocumentTemplate):
            entry_query = entry_query.pk
        mapped_entry, entry_fields = map_entry_with_dots(entry, fields)
        mapped_query[mapped_entry] = map_query(entry_query, entry_fields)
    return mapped_query
```

In addition, may I propose accepting **kwargs in the `find` and `find_one` function as the search parameters, allowing the following method call to exist?

```python
Foo.find(bar="barid")
```

I've looked into the source, and I admit, I'm not entirely sure what that would look like in terms of a PR. On the other hand, it would be marginally more pythonic, but would stray away from Mongo conventions.

I am more than happy to PR both of these, and plan on implementing my own fix for the first internally tonight.

**Comment by lafrech** [comment]
At a quick glance, the fix seems reasonable.

What about `EmbeddedDocuments`? Should we dump them to pass them in the query? I haven't been playing with raw mongo for a while.

Does it make sense to query like this?

    find({'pet': {'age': 12, 'name': 'Toby'}}

In this case, if I have an embedded Pet object:

    pet = Pet(age=12, name='Toby')

and I query

    find({'pet': pet}

I'll get the same `Cannot encode object: <object>`, right?

Shall we dump the embedded Pet document in the query mapper?

Sorry if I'm off-topic.

I have no strong opinion about the `**kwargs` thing yet.

**Comment by enugentdt** [comment]
That's a good point. My _guess_ would be that the best course would be to only grab the PK if it's a Document, and not an EmbeddedDocument. Though, Python's `isinstance` and I have a bad relationship. I'll go searching for a good way to tell if something is embedded vs not embedded.

And with `**kwargs`, it's mostly a QoL thing that I enjoyed from Mongoengine. It's not super important to me - I wrote wrappers around the `find` and `find_one` functions anyway, where I just dump kwargs.

**Comment by lafrech** [comment]
Using `isinstance` is fine as they don't inherit from each other. I was just thinking we could be tempted to fix the `EmbeddedDocument` case while we're at it.

@touilleMan tried to keep umongo as simple as possible, avoiding hidden automatic behaviours. Maybe we're going too far and the user should be responsible for passing a pk and not an object to the find method. From [the docs](https://umongo.readthedocs.io/en/latest/index.html):

> Stay close to the standards MongoDB driver to keep the same API when possible: use `find({"field": "value"}`) like usual but retrieve your data nicely OO wrapped !

Manually grabbing the pk from the object is consistent with the rest of the API. Like the need to call fetch rather than automatically dereference, for instance. My concern is that we may be opening the door to all sorts of similar issues if we start to support passing anything else than raw Mongo queries.

Just thinking out loud, no strong objection. After all, we already do the field name mapping. And I don't see what the fix you're proposing could break.

**Comment by enugentdt** [comment]
It's all a game of balance.

Honestly, in my mind, I could make arguments either way. One thing this _does_ help is the idea of an object as a reference, rather than a key as a reference. Unfortunately, Mongo doesn't do "true" references (only embedded objects). However, doing this reinforces the idea that an object is the final reference.

Of course, playing devil's advocate, one could reply that it's important to stay as close to the root Mongo API as possible, as that is the purpose of this library. If that's the general consensus, I somewhat wonder if it would make sense to put another library on top of this, that just stuffs QoL features in to make learning easier. But hey, I'm getting ahead of myself.

But the note on "opening the door" to hosts of other issues is true. For example, I use multiple circular references (not that I'm proud of it) in my production database. Auto dereference would not only slow down my queries, but also crash the servers.

Not that I have any real say, but would the idea of "quality of life, but no speed/query sacrifice" be a good route to take in this case? I.e. make things easier without making them more expensive - ergo, the whole circular reference issue.

**Comment by lafrech** [comment]
Note the `EmbeddedDocument` use case seems broken anyway due to #55.

**Comment by lafrech** [comment]
Let's not add the **kwargs.

The issue in OP should be solved by #303. Thanks for reporting.
